### PR TITLE
Ldap authentication

### PR DESCRIPTION
**Pull request by cawallin**

### CODE REVIEW

File: prestoadmin/prestoclient.py [hunk 1]
```diff
@@ -129,10 +137,12 @@ def execute_query(self, sql, schema="default", catalog="hive"):
 
             self.response_from_server = json.loads(answer)
             _LOGGER.info("Query executed successfully: %s" % (sql))
+            _LOGGER.info("Using LDAP = %s" % self.coordinator_config.use_ldap())
```
Comment (ebd2) [inline prestoadmin/prestoclient.py]: I propose moving this above where we make the request and including the username. LIke maybe around 124?

File: prestoadmin/prestoclient.py [hunk 2]
```diff
@@ -129,10 +137,12 @@ def execute_query(self, sql, schema="default", catalog="hive"):
...
             return True
         except (HTTPException, socket.error) as e:
             _LOGGER.error("Error connecting to presto server at: " +
                           self.server + ":" + str(self.port) + ' ' + e.message)
+            _LOGGER.info("Using LDAP = %s" % self.coordinator_config.use_ldap())
```
Comment (ebd2) [inline prestoadmin/prestoclient.py]: Which would make this redundant?

File: prestoadmin/prestoclient.py [hunk 3]
```diff
@@ -129,10 +137,12 @@ def execute_query(self, sql, schema="default", catalog="hive"):
...
             return False
         except ValueError as e:
             _LOGGER.error('Error connecting to Presto server: ' + e.message +
```
Comment (ebd2) [inline prestoadmin/prestoclient.py]: If the one on the HTTPException isn't redundant, why not log here too?

File: prestoadmin/prestoclient.py [hunk 4]
```diff
@@ -153,9 +163,14 @@ def get_response_from(self, uri):
         parts[0] = None
         parts[1] = None
         location = urlparse.urlunsplit(parts)
-
         conn = self._get_connection()
-        conn.request("GET", location)
+        headers = {"X-Presto-User": self.user}
+        if self.coordinator_config.use_ldap():
```
Comment (ebd2) [inline prestoadmin/prestoclient.py]: Extract to a function `_add_auth_headers(headers)`?

File: prestoadmin/util/presto_config.py [hunk 5]
```diff
@@ -96,3 +101,14 @@ def get_https_port(self):
...
     def get_http_port(self):
         return int(self._lookup(HTTP_PORT_KEY))
+
+    def use_ldap(self):
+        if AUTHENTICATION_KEY in self.config_properties:
+            return self.config_properties[AUTHENTICATION_KEY] == 'LDAP'
+        return False
+
+    def get_ldap_user(self):
+        return self.config_properties[LDAP_CLIENT_USER_KEY]
```
Comment (ebd2) [inline prestoadmin/util/presto_config.py]: Use _lookup()

File: prestoadmin/util/presto_config.py [hunk 6]
```diff
@@ -28,6 +28,9 @@
 HTTPS_PORT_KEY = 'http-server.https.port'
 CLIENT_KEYSTORE_PATH_KEY = 'internal-communication.https.keystore.path'
 CLIENT_KEYSTORE_PASSWORD_KEY = 'internal-communication.https.keystore.key'
+AUTHENTICATION_KEY = 'http-server.authentication.type'
+LDAP_CLIENT_USER_KEY = 'http-client.authentication.ldap.user'
```
Comment (cawallin) [inline prestoadmin/util/presto_config.py]: changed it to internal-communication

File: prestoadmin/util/presto_config.py [hunk 7]
```diff
@@ -96,3 +101,14 @@ def get_https_port(self):
...
+    def use_ldap(self):
+        if AUTHENTICATION_KEY in self.config_properties:
+            return self.config_properties[AUTHENTICATION_KEY] == 'LDAP'
+        return False
+
+    def get_ldap_user(self):
+        return self.config_properties[LDAP_CLIENT_USER_KEY]
+
+    def get_ldap_password(self):
+        return self.config_properties[LDAP_CLIENT_PASSWORD_KEY]
```
Comment (ebd2) [inline prestoadmin/util/presto_config.py]: ditto

File: prestoadmin/prestoclient.py [hunk 8]
```diff
@@ -310,3 +325,21 @@ def _get_private_key(self, keystore):
             error('No alias %s found in %s. Available aliases are %s' %
                   (alias, self.coordinator_config.get_client_keystore_path(),
                    all_keys))
+
+    @staticmethod
+    def create_auth_headers(user, password):
+        if not user:
```
Comment (ebd2) [inline prestoadmin/prestoclient.py]: Redundant if you use _lookup in presto_config, unless you're envisioning some other path that calls this that could screw it up.

File: prestoadmin/prestoclient.py [hunk 9]
```diff
@@ -310,3 +325,21 @@ def _get_private_key(self, keystore):
...
+            error('LDAP user (taken from %s in %s on the coordinator) cannot be null or empty' %
+                  (LDAP_CLIENT_USER_KEY, os.path.join(REMOTE_CONF_DIR, CONFIG_PROPERTIES)))
+            return {}
+        if not password:
+            error('LDAP password (taken from %s in %s on the coordinator) cannot be null or empty' %
+                  (LDAP_CLIENT_PASSWORD_KEY, os.path.join(REMOTE_CONF_DIR, CONFIG_PROPERTIES)))
+            return {}
+
+        if ':' in user:
+            error("LDAP user cannot contain ':': %s" % user)
```
Comment (ebd2) [inline prestoadmin/prestoclient.py]: why check `:` specifically? 💩 probably isn't a valid character either.

File: tests/unit/test_prestoclient.py [hunk 10]
```diff
@@ -177,3 +177,35 @@ def test_execute_query_get_port(self, sudo_mock, conn_mock, mock_presto_config):
         self.assertEqual(client.rows, [])
         self.assertEqual(client.next_uri, '')
         self.assertEqual(client.response_from_server, {})
+
+    def test_create_authorization_headers(self, mock_presto_config):
+        auth_headers = PrestoClient.create_auth_headers("Aladdin", "open sesame")
+        expected_auth_headers = {"Authorization": "Basic QWxhZGRpbjpvcGVuIHNlc2FtZQ=="}
+        self.assertEqual(auth_headers, expected_auth_headers)
+
+    @patch('prestoadmin.prestoclient.error')
```
Comment (ebd2) [inline tests/unit/test_prestoclient.py]: I'd argue that the negative tests are all redundant, based on my comments above.

File: prestoadmin/prestoclient.py (hunk 8 above)
Comment (cawallin) [inline prestoadmin/prestoclient.py]: eh, I think it's good to keep these generic checks to verify that the LDAP contract is fulfilled

File: prestoadmin/prestoclient.py (hunk 9 above)
Comment (cawallin) [inline prestoadmin/prestoclient.py]: from @arhimondr -- "Check that user and password are not empty, and that username does not contain the : character."

**Comment by cawallin** [comment]
@ebd2 updated

File: prestoadmin/prestoclient.py (hunk 9 above)
Comment (anusudarsan) [inline prestoadmin/prestoclient.py]: Right https://tools.ietf.org/html/rfc2617#section-2

Comment (ebd2) [inline prestoadmin/prestoclient.py]: If it's in the RFC, that's fine with me.

**Pull request merged**